### PR TITLE
update submodules

### PR DESCRIPTION
**Pull request by zhaozg**
- ce5f29b bump version to v1.49.0
- 6fbcb52 bump libuv to v1.49.0
- c9a472c bump lua-compat-5.3 to v0.14.3
- 67ade8d bump LuaJIT to 2.1.1727870382

new API list:
- uv.utf16_length_as_wtf8
- uv.utf16_to_wtf8
- uv.wtf8_length_as_utf16
- uv.wtf8_to_utf16

### CODE REVIEW

File: docs.md [hunk 1]
```diff
@@ -4051,6 +4051,58 @@ metrics counters.
...
 
+## String manipulation functions
+
+These string utilities are needed internally for dealing with Windows, and are exported to allow clients to work uniformly with this data when the libuv API is not complete.
+
+**Note**:
+
+1. New in luv version 1.49.0.
+2. A UTF-16 character is 2 bytes, and a UTF-8 character is 1 byte.
+3. All input and returned utf8 string or utf16 string should be NULL-terminated or these API.
```
Comment (rdw-software) [inline docs.md:4062]:
```suggestion
3. All inputs and return values (UTF-8 or UTF-16 strings) should be NUL terminated.
```

File: docs.md [hunk 2]
```diff
@@ -4051,6 +4051,58 @@ metrics counters.
 - `events` : `integer`
 - `events_waiting` : `integer`
 
+## String manipulation functions
+
+These string utilities are needed internally for dealing with Windows, and are exported to allow clients to work uniformly with this data when the libuv API is not complete.
+
+**Note**:
```
Comment (rdw-software) [inline docs.md:4058]:
```suggestion
**Notes**:
```

File: tests/test-misc.lua [hunk 3]
```diff
@@ -219,4 +219,16 @@ return require('lib/tap')(function (test)
     end
   end, "1.45.0")
 
+  test("uv.wtf8 and utf8 conversion", function(print, p, expect, uv)
+    -- default encode is utf8/wtf8
```
Comment (rdw-software) [inline tests/test-misc.lua:223]:
```suggestion
    -- default encoding is utf8/wtf8
```

File: src/misc.c [hunk 4]
```diff
@@ -782,3 +782,62 @@ static int luv_clock_gettime(lua_State* L) {
...
+  return ret;
+}
+
+static int luv_wtf8_length_as_utf16(lua_State *L) {
+  size_t sz;
+  ssize_t ssz;
+  const char* utf8 = luaL_checklstring(L, 1, &sz);
+  luaL_argcheck (L, sz > 1 && utf8[sz-1] == '\0',
+                 1, "expected NULL-terminated string");
+  ssz = uv_wtf8_length_as_utf16(utf8);
```
Comment (squeek502) [inline src/misc.c:823]: This doesn't seem like the Lua way to handle this. Maybe something like this instead:

```suggestion
  const char* utf8 = luaL_checklstring(L, 1, &sz);
  char* utf8terminated = malloc(sz + 1);
  if (utf8terminated == NULL) return luaL_error(L, "failed to allocate %zu bytes", sz + 1);
  memcpy(utf8terminated, utf8, sz);
  utf8terminated[sz] = 0;
  ssz = uv_wtf8_length_as_utf16(utf8terminated);
  free(utf8terminated);
```

And then the same thing for `luv_wtf8_to_utf16`

**Comment by zhaozg** [comment]
Addressed

**Review by squeek502 — APPROVED** [review]
Added a commit with some changes. If you like the changes I made, then feel free to merge this.

**Comment by zhaozg** [comment]
> Some cleanup/improvements to wtf8/utf16 functions

I really like it.

File: src/misc.c [hunk 5]
```diff
@@ -782,3 +782,72 @@ static int luv_clock_gettime(lua_State* L) {
...
 #endif
+
+#if LUV_UV_VERSION_GEQ(1, 49, 0)
+
+static int luv_utf16_length_as_wtf8(lua_State* L) {
+  size_t sz;
+  const uint16_t *utf16 = (const uint16_t *)luaL_checklstring(L, 1, &sz);
+  ssize_t utf16_len = sz/2;
+  sz = uv_utf16_length_as_wtf8(utf16, utf16_len);
+  /* The returned length includes a NUL terminator, but we use Lua style string */
```
Comment (truemedian) [inline src/misc.c:793]: This comment is incorrect and does not represent the code below it. The `sz` calculated does not include any NUL terminator.

File: src/misc.c [hunk 6]
```diff
@@ -782,3 +782,72 @@ static int luv_clock_gettime(lua_State* L) {
...
+static int luv_wtf8_length_as_utf16(lua_State *L) {
+  size_t sz;
+  ssize_t ssz;
+  const char* utf8 = luaL_checklstring(L, 1, &sz);
+  char *s = malloc(sz+1);
+  if (s == NULL) return luaL_error(L, "failed to allocate %zu bytes", sz + 1);
+  memcpy(s, utf8, sz);
+  s[sz] = '\0';
+  ssz = uv_wtf8_length_as_utf16(s);
+  free(s);
```
Comment (truemedian) [inline src/misc.c:827]:
```suggestion
  const char* wtf8 = luaL_checkstring(L, 1);
  ssize_t ssz = uv_wtf8_length_as_utf16(wtf8);
```

This copy of the input string is unnecessary, `checkstring` is guaranteed to give us a NUL terminated string.

File: src/misc.c [hunk 7]
```diff
@@ -782,3 +782,72 @@ static int luv_clock_gettime(lua_State* L) {
...
+  memcpy(s, utf8, sz);
+  s[sz] = '\0';
+  ssz = uv_wtf8_length_as_utf16(s);
+  utf16 = malloc(ssz * 2);
+  if (utf16 == NULL) return luaL_error(L, "failed to allocate %zu bytes", ssz * 2);
+  uv_wtf8_to_utf16(s, utf16, ssz);
+  /* The returned string includes a NUL terminator, but we use Lua style string */
+  lua_pushlstring(L, (const char*)utf16, (ssz-1) * 2);
+  free(utf16);
+  free(s);
```
Comment (truemedian) [inline src/misc.c:849]:
```suggestion
  const char *wtf8 = luaL_checkstring(L, 1);
  ssize_t ssz = uv_wtf8_length_as_utf16(s);
  uint16_t *utf16 = malloc(ssz * 2);
  if (utf16 == NULL) return luaL_error(L, "failed to allocate %zu bytes", ssz * 2);
  uv_wtf8_to_utf16(wtf8, utf16, ssz);
  /* The returned string includes a NUL terminator, but we use Lua style string */
  lua_pushlstring(L, (const char*)utf16, (ssz-1) * 2);
  free(utf16);
```

Another unnecessary copy of the input string, `checkstring` is guaranteed to give us a NUL terminated string.

File: src/misc.c [hunk 8]
```diff
@@ -782,3 +782,72 @@ static int luv_clock_gettime(lua_State* L) {
   return 1;
 }
 #endif
+
+#if LUV_UV_VERSION_GEQ(1, 49, 0)
```
Comment (truemedian) [inline src/misc.c:786]: This is suspicious, Libuv documentation lists these as being introduced in 1.47.0

**Review by truemedian — COMMENTED** [review]
Just noticed this was merged, and these don't necessarily require blocking this merge but there are still a few things that should have been caught here.

Especially relating to what version of libuv these functions were introduced.

File: src/misc.c (hunk 8 above)
Comment (squeek502) [reply src/misc.c:786]: They were added to `uv.h` in 1.47.0, but were not exported until 1.49.0: https://github.com/libuv/libuv/pull/4437

Comment (zhaozg) [reply src/misc.c:786]: These API introduced in 1.47.0, but document and public in uv.h v1.49.0, Will compile fail if we set 1.47.0.

**Pull request merged**